eventuals/os.h

```diff
@@ -60,6 +67,153 @@ inline void SetAffinity(std::thread& thread, const size_t cpu) {
 }
 #endif // __MACH__
 
+#if defined(__linux__) || defined(__MACH__)
+
+namespace details {
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Lambda>
+void* call_target(void* target) {
+  Lambda* lambda = reinterpret_cast<Lambda*>(target);
+  (*lambda)();
+  delete lambda;
+  return nullptr;
+}
+
+////////////////////////////////////////////////////////////////////////
+
+} // namespace details
+
+class Thread {
+ public:
+  Thread()
+    : thread_handle_{pthread_t{}},
+      joinable_(false) {}
+
+  ~Thread() noexcept(false) {
+    if (joinable_) {
+      throw std::logic_error{"A thread was left not joined/not detached"};
+    }
+  }
+
+  Thread(const Thread& other) = delete;
+
+  // IMPORTANT NOTE: on macos stacksize should be a multiple of the system
+  // page size!!!
+  // (check_line_length skip)
+  // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_attr_setstacksize.3.html
+  template <typename Callable>
+  Thread(Callable&& callable, const size_t stack_size_in_bytes = 8'388'608)
+    : joinable_{true} {
+    CHECK_GE(stack_size_in_bytes, PTHREAD_STACK_MIN)
+        << "Stack size should not be less than"
+           " the system-defined minimum size";
+
+    State_<Callable>* state =
+        new State_<Callable>{std::forward<Callable>(callable)};
+
+    using lambda_type = std::remove_reference_t<decltype(*state)>;
+
+    pthread_attr_t attr;
+
+    CHECK_EQ(0, pthread_attr_init(&attr))
+        << "Error in pthread_attr_init(...)";
+
+    // IMPORTANT NOTE: on macos stacksize should be a multiple of the system
+    // page size, otherwise `pthread_attr_setstacksize` will fail.
+    // (check_line_length skip)
+    // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_attr_setstacksize.3.html
+    CHECK_EQ(0, pthread_attr_setstacksize(&attr, stack_size_in_bytes))
+        << "Error in pthread_attr_setstacksize(...), "
+           "if you are on macOS - probably you are trying"
+           " to set the stack size which is not a multiple"
+           " of the system page size";
+
+    int creation_result = pthread_create(
+        &thread_handle_,
+        &attr,
+        &details::call_target<lambda_type>, // void* (*pointer) (void*)
+        state);
+
+    CHECK_EQ(0, pthread_attr_destroy(&attr))
+        << "Error in pthread_attr_destroy(...)";
+
+    if (creation_result != 0) {
+      delete state;
+      throw std::runtime_error{
+          "Failed to create a new thread with "
+          "'pthread_create(), error code = '"
+          + std::to_string(creation_result)};
+    }
+  }
```

Comment:
```suggestion
  template <typename Callable>
  Thread(const std::string& name, Callable&& callable, const size_t stack_size_in_bytes = 8'388'608)
    : joinable_{true} {
    CHECK_GE(stack_size_in_bytes, PTHREAD_STACK_MIN)
        << "Stack size should not be less than"
           " the system-defined minimum size";

    pthread_attr_t attr;

    PCHECK(pthread_attr_init(&attr) == 0);
    
     // TODO(artur): use pthread_attr_setname_np

    // IMPORTANT NOTE: on macos stacksize should be a multiple of the system
    // page size, otherwise `pthread_attr_setstacksize` will fail.
    // (check_line_length skip)
    // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_attr_setstacksize.3.html
    PCHECK(pthread_attr_setstacksize(&attr, stack_size_in_bytes) == 0)
        << "Failed to set the stack size via 'pthread_attr_setstacksize' (if you are on macOS - probably you are trying"
         << " to set the stack size which is not a multiple"
         << " of the system page size)";

    PCHECK(pthread_create(
        &thread_handle_,
        &attr,
        [](void* arg) {
          Callable* callable = reinterpret_cast<Callable*>(arg);
          try {
            (*callable)();
          } catch (const std::exception& e) {
            const std::string& name = GetThreadName();
            LOG(WARNING) << "caught exception while running thread '" << name << "': " << e.what();
          } catch (...) {
            const std::string& name = GetThreadName();
            LOG(WARNING) << "caught unknown exception while running thread '" << name << "'";
          }
          delete callable; 
        }
        new Callable(std::forward<Callable>(callable)) == 0)
        << "Failed to create a new thread via 'pthread_create'";

    CHECK_EQ(0, pthread_attr_destroy(&attr));
  }
```